File: frontend/src/app/core/utils/purchase-bill-calculator.ts
/**
 * Mirrors PurchaseBill.Application.Common.PurchaseBillCalculator on the backend, so the grid
 * shows live totals as the user types instead of waiting on a round trip. The backend
 * recomputes the same values when the bill is saved - it is the source of truth, this is only
 * for instant feedback.
 */

function round2(value: number): number {
  return Math.round((value + Number.EPSILON) * 100) / 100;
}

export function calculateMargin(standardCost: number, standardPrice: number): number {
  return round2(standardPrice - standardCost);
}

export function calculateTotalCost(
  standardCost: number,
  quantity: number,
  discountPercent: number,
): number {
  return round2(standardCost * quantity * (1 - discountPercent / 100));
}

export function calculateTotalSelling(standardPrice: number, quantity: number): number {
  return round2(standardPrice * quantity);
}

File: backend/src/PurchaseBill.Application/Common/PurchaseBillCalculator.cs
namespace PurchaseBill.Application.Common;

/// <summary>
/// Line-item calculations for the Purchase Bill form, per the assignment brief:
///   Total Cost    = Standard Cost * Quantity * (1 - Discount%)
///   Total Selling = Standard Price * Quantity
///   Margin        = Standard Price - Standard Cost
///
/// Kept here (rather than inline in the service) so the exact same rule can be unit tested and
/// referenced from one place; the Angular side mirrors this for live, pre-submit totals.
/// </summary>
public static class PurchaseBillCalculator
{
    public static decimal Margin(decimal standardCost, decimal standardPrice) =>
        Round(standardPrice - standardCost);

    public static decimal TotalCost(decimal standardCost, decimal quantity, decimal discountPercent) =>
        Round(standardCost * quantity * (1 - discountPercent / 100m));

    public static decimal TotalSelling(decimal standardPrice, decimal quantity) =>
        Round(standardPrice * quantity);

    private static decimal Round(decimal value) => Math.Round(value, 2, MidpointRounding.AwayFromZero);
}
